lib/chef/knife/windows_cert_generate.rb

```diff
@@ -0,0 +1,151 @@
+# Author:: Mukta Aphale (<mukta.aphale@clogeny.com>)
+# Copyright:: Copyright (c) 2014-2016 Chef Software, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative "../knife"
+
+class Chef
+  class Knife
+    class WindowsCertGenerate < Knife
+
+      attr_accessor :thumbprint, :hostname
+
+      banner "knife windows cert generate FILE_PATH (options)"
+
+      deps do
+        require "openssl"
+        require "socket"
+      end
+
+      option :hostname,
+        short: "-H HOSTNAME",
+        long: "--hostname HOSTNAME",
+        description: "Use to specify the hostname for the listener.
+        For example, --hostname something.mydomain.com or *.mydomain.com.",
+        required: true
+
+      option :output_file,
+        short: "-o PATH",
+        long: "--output-file PATH",
+        description: "Specifies the file path at which to generate the 3 certificate files of type .pfx, .b64, and .pem. The default is './winrmcert'.",
+        default: "winrmcert"
+
+      option :key_length,
+        short: "-k LENGTH",
+        long: "--key-length LENGTH",
+        description: "Specifies the key length of the certificate. The default is 2048",
+        default: "2048"
+
+      option :cert_validity,
+        short: "-cv MONTHS",
+        long: "--cert-validity MONTHS",
+        description: "Specifies how long the certificate will be valid. The default is 24 months",
+        default: "24"
+
+      option :cert_passphrase,
+        short: "-cp PASSWORD",
+        long: "--cert-passphrase PASSWORD",
+        description: "Specifies certificate's passphrase."
+
+      def generate_keypair
+        OpenSSL::PKey::RSA.new(config[:key_length].to_i)
+      end
+
+      def prompt_for_passphrase
+        passphrase = ""
+        loop do
+          print "Passphrases do not match. Try again.\n" unless passphrase.empty?
+          passphrase = ui.ask("Enter certificate passphrase (empty for no passphrase):", echo: false)
+          confirm_passphrase = ui.ask("Confirm the passphrase:", echo: false)
+          break if passphrase == confirm_passphrase
+        end
+      end
```

Comment:
```suggestion
      def prompt_for_passphrase
        loop do
          passphrase = ui.ask("Enter certificate passphrase (empty for no passphrase):", echo: false)
          confirm_passphrase = ui.ask("Confirm the passphrase:", echo: false)
          return passphrase if passphrase == confirm_passphrase
          puts "Passphrases do not match. Try again."
        end
      end
```

Is this function currently doing anything? Plus code cleanup.
```
irb(main):014:0> def foo(arg) ; loop do ; print "wat? " ; input = gets ; break if input.strip == arg ; end ; end
=> :foo
irb(main):015:0> foo("lunch")
wat? dinner
wat? hello
wat? lunch
=> nil
```

Reply:
It successfully prompted for password when I tested it locally and threw and error and prompted again when they didn't match.

Reply:
I don't really understand how this could work, but I can definitely understand not wanting to change it unnecessarily. 👍

Reply:
With matching passwords:

```
Do you really want to overwrite existing certificates? (Y/N) y
Enter certificate passphrase (empty for no passphrase):
Confirm the passphrase:
Generated Certificates:
- winrmcert.pfx - PKCS12 format key pair. Contains public and private keys, can be used with an SSL server.
- winrmcert.b64 - Base64 encoded PKCS12 key pair. Contains public and private keys, used by some cloud provider API's to configure SSL servers.
- winrmcert.pem - Base64 encoded public certificate only. Required by the client to connect to the server.
Certificate Thumbprint: D8669828883C1F1FF5BCFA22B9B9D5C8582A1E84
```

When passwords don't match the first time:

```
Do you really want to overwrite existing certificates? (Y/N) y
Enter certificate passphrase (empty for no passphrase):
Confirm the passphrase:
Passphrases do not match. Try again.
Enter certificate passphrase (empty for no passphrase):
Confirm the passphrase:
Generated Certificates:
- winrmcert.pfx - PKCS12 format key pair. Contains public and private keys, can be used with an SSL server.
- winrmcert.b64 - Base64 encoded PKCS12 key pair. Contains public and private keys, used by some cloud provider API's to configure SSL servers.
- winrmcert.pem - Base64 encoded public certificate only. Required by the client to connect to the server.
Certificate Thumbprint: EC5043FBDD3484C3DD33D7BDAB8C8DE922780D60
```

Reply:
I cleaned this one up though to avoid setting the empty string